protocols/gossipsub/src/lib.rs

```diff
@@ -156,18 +156,73 @@ mod types;
 
 mod rpc_proto;
 
-pub use self::behaviour::{Gossipsub, GossipsubEvent, MessageAuthenticity};
+pub use self::behaviour::{Behaviour, Event, MessageAuthenticity};
 pub use self::transform::{DataTransform, IdentityTransform};
 
-pub use self::config::{GossipsubConfig, GossipsubConfigBuilder, GossipsubVersion, ValidationMode};
 pub use self::peer_score::{
     score_parameter_decay, score_parameter_decay_with_base, PeerScoreParams, PeerScoreThresholds,
     TopicScoreParams,
 };
 pub use self::topic::{Hasher, Topic, TopicHash};
-pub use self::types::{
-    FastMessageId, GossipsubMessage, GossipsubRpc, MessageAcceptance, MessageId,
-    RawGossipsubMessage,
-};
+pub use self::types::{FastMessageId, Message, MessageAcceptance, MessageId, RawMessage, Rpc};
+
+pub use self::config::{Config, ConfigBuilder, ValidationMode, Version};
+
+pub use self::error::{HandlerError, PublishError, SubscriptionError, ValidationError};
+
+#[deprecated(
+    since = "0.44.0",
+    note = "Use re-exports that omit `Gossipsub` prefix, i.e. `libp2p::gossipsub::HandlerError"
+)]
+pub type GossipsubHandlerError = HandlerError;
```

Comment:
It is slightly related but it would be really great if we can get rid of the public `error` module while we are at it. I'd prefer to not have any public modules actually. They are an easy semver hazard because types that shouldnt' be part of the public API easily slip into it.

To achieve this in a backwards-compatible way, we need to:

1. Make the `error` module private and rename it to `error_priv`
2. Make a new, inline `error` module that contains type aliases with deprecation notice, re-exporting the errors from `error_priv`
3. Export all errors in the root module (you are already doing this on line 171)

This allows us to migrate users towards importing the errors from the root module and we can remove the public `error` module in a future release!

See what we did in this PR for example: https://github.com/libp2p/rust-libp2p/pull/3238/files#diff-acd49d828b2b4f41554331034dd696f7946d8c1becc7635d118435454c262cde

Everything used to be under a `v2` module that now only contains deprecated type-aliases that re-export symbols from their new location. Some of the new locations are _crate-private_ however. In other words, we will be making certain types inaccessible in a future release. That is because they should have never been part of the public API.

I see some opportunities for this in gossipsub too:

- `protocol` module
- `metrics` (contains many public types and functions that a user doesn't need to interact with)
- `subscription_filter` (can be completely flattened, i.e. re-export from root to shorten import paths)
- `time_cache` (not sure why that is public at all?)

This is quite an undertaking by itself so it is probably best done in a separate PR. Let me know if you are up for it :)

Reply:
Perfectly understandable. I like this direction. I'm not sure about making `V2` APIs in these kind of changes, but let's talk about it in the next PR ;)

Reply:
I've created a new issue to track these changes -> https://github.com/libp2p/rust-libp2p/issues/3392